KEP-5233: proposal for NodeReadinessGates (#5416)
kubernetes/enhancements: keps/sig-node/5233-node-readiness-gates/README.md

> # KEP-5233: Node Readiness Gates

Reviewer: If you are hoping to merge this for 1.35, please request a PRR review. You will need to add prod-readiness file and find an approver for that.

## Proposal

### Main Idea

This proposal's core idea is to make a node's transition to a fully schedulable state dependent not only on the Kubelet's default `Ready` condition but also on custom criteria defined by the cluster administrator.

The mechanism works as follows:

1. A Node is created with 'readiness-probes' configuration and well-defined taint(s) for holding node-scheduling - referred in this KEP as __gates__.
2. The readiness status of essential components (like CNI plugins or CSI/DRA drivers) is reported as `Conditions` in the `NodeStatus`.
3. These `Conditions` are updated by reliable, node-local entities (such as Node-Problem-Detector or Kubelet itself), which probe the configured components' health-endpoints and update the `/status` sub-resource.
4. This allows a controller to gate the node's schedulability. The node is considered fully available for pod scheduling only when these custom readiness `Conditions` are met.

This approach allows critical components to directly influence when a node is ready, complementing the existing `Ready` condition with more granular, user-defined control.

> ### User Stories (Optional)

Reviewer 2: ```suggestion
### User Stories
```

## Design Details

### High Level Design:

```mermaid
sequenceDiagram
    participant NC as Node Controller
    participant NA as Node-Agent (probes)
    participant N as Node
    participant CNI as CNI Component
    participant DS as DaemonSet Controller
    participant S as Scheduler
    participant P as Pod (Regular)
    participant DP as Pod (DaemonSet)

    Note over N: Node Bootstrap Phase
    N->>N: Configure spec.readinessGates<br/>[network.k8s.io/CNIReady]
    NC->>N: Add taint: node.kubernetes.io/not-ready:NoSchedule
    N->>N: Node marked as Ready after Kubelet and Runtime are ready.<br/>This does not signal ready for workload scheduling.

    Note over CNI,DS: Component Initialization
    DS->>DP: Create DaemonSet pods with tolerations
    Note over DP: Tolerates node.kubernetes.io/not-ready taint
    S->>DP: Schedule DaemonSet pods (can run on tainted node)
    CNI->>CNI: Initialize CNI components

    Note over NA, CNI: Node-Agent Probes for Readiness
    NA->>CNI: Probe for readiness (e.g., check health endpoint)
    CNI-->>NA: Report Ready

>     NA->>N: Patch status.conditions:<br/>network.k8s.io/CNIReady=True

Reviewer 2: We shouldn't make an assumption that network plugins use CNI.

-  Add an optional field `readinessGates` to `NodeSpec`.
-  Type: `[]NodeReadinessGate`
-  `NodeReadinessGate` struct

```go
// NodeReadinessGate specifies a condition that  must be true for the node to be considered fully-schedulable.
type NodeReadinessGate {
  // ConditionType refers to a condition in the Node's `status.Condition` array with matching type.
  // Each conditionType must be unique within node.spec.readinessGates
  // +required
  ConditionType v1.NodeConditionType `json:”conditionType”`
}

// NodeSpec describes the attributes that a node is created with
type NodeSpec struct {
  // existing fields..

  // ReadinessGates give declared readiness-requirements for node. 
  // This field is immutable. It can only be set during node creation.
  //
  // +featureGate=NodeReadinessGates
  // +optional
  ReadinessGates []NodeReadinessGate `json:"readinessGates,omitempty" patchStrategy:"merge" patchMergeKey:"conditionType"`
}
```

Example Node Manifest:

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gpu-node-1
spec:
  # ... existing fields
  readinessGates:
    # A list of condition types that must be 'True' for the node to be considered ready.

>     - conditionType: "vendor.com/DeviceDriverReady"
>     - conditionType: "network.k8s.io/CalicoCNIReady"

Reviewer 2: ```suggestion
    - conditionType: "vendor.example.com/DeviceDriverReady"
    - conditionType: "vendor.example/NetworkReady"
```

Kubelet will dynamically manage these labels during node-status synchronization. Kubelet will detect the configured readiness gates and ensure corresponding labels are set and cleaned up if necessary. This ensures advertised node readiness-gates are accurate throughout the node-lifecycle.

### Probing Mechanisms (Optional)

To support `readinessGates`, there could be a privileged node-local entity to probe the component readiness continuously. Based on the probe results, the node-agent will manage the lifecycle of the corresponding `conditionType` in the node's status, patching it when there is a status change after success / failure threshold is met.

#### Kubelet Local Probing

One option is to extend Kubelet with a new, extensible probing framework to evaluate probes and publis matching NodeConditions. This design could leverage existing `v1.Probe` struct for node-readiness configuration.

**Kubelet Configuration Example**

```yaml
# Hypothetical Kubelet Configuration
nodeReadinessProbes:
  - name: "CNIReady"

>     conditionType: "network.k8s.io/CalicoCNIReady"

Reviewer 2: ```suggestion
    conditionType: "vendor.example/NetworkReady"
```

    probe:
      httpGet:
        host: "127.0.0.1"
        port: 9090
        path: "/healthz"
    periodSeconds: 15
    failureThreshold: 3
  - name: "StorageReady"

>     conditionType: "storage.k8s.io/csi/abc-driver/Ready"

Reviewer 2: ```suggestion
    conditionType: "abc.other-vendor.example/Ready"
```

* **Beta:**
    * Feature gate default `true`.
    * e2e tests implemented and passing consistently.
    * Scalability and performance testing performed.
    *

* **GA:**
    *

### Upgrade / Downgrade Strategy

- **Upgrade (Gate Disabled -> Enabled):** Existing nodes will not have `spec.readinessGates` defined. They will continue to function as before. New nodes or existing nodes where the field is added will start enforcing the gates. Probing plugins / configuration need deploying/updating to manage the conditions.
- **Downgrade (Gate Enabled -> Disabled):** The node-controller will stop evaluating the `readinessGates` field. Nodes previously blocked by gates may become schedulable (based only on `Ready` condition and taints). The `readinessGates` field remains in the spec but has no effect. Node plugins patching conditions will continue but wont affect scheduling; can still be useful for observability.

### Version Skew Strategy

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

1. Feature gate (also fill in values in `kep.yaml`)

>     -  Feature gate name:

Reviewer 2: Mention `NodeReadinessGates` here

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

## Drawbacks

* Adds complexity to the node readiness definition.
* Requires plugins / configurations to be written correctly to manage the status conditions. Misconfigured probes could render nodes perpetually unschedulable.
* Potential for increased `node.status` patch requests to the API server, although targeted patching of conditions should be manageable.
* Defining the right set of gates requires careful consideration by the cluster administrator.

## Alternatives

> 

Reviewer 2: For autoscaling, cluster autoscalers can directly pay attention to the existing `.status.conditions` on nodes. I think that's a viable alternative and one we should list.

### Relying Solely On Retries (Eventual Consistency)

Like any distributed system, one could simply allow failures and rely on retries until operations eventually succeed. However, this reactive approach stands in contrast to the proactive management offered by Node Readiness Gates. A proactive system avoids cascading failures, a risk particularly in large-scale workloads like ML training jobs. This makes readiness-gates a critical guardrail for ensuring that complex dependencies are met before scheduling begins.

> ### Initial Taints without a Central API
> 
> This approach uses `--register-with-taints` to apply multiple readiness taints at startup. Each component is then responsible for removing its own taint. This is less flexible and discoverable than a formal, versioned API for defining the readiness requirements. In addition,  due to operational complexity where every critical daemonset needs to be tolerating every other potential readiness taint possible. This is unmanageable in a practical scenario where the components could be managed by different teams / providers. 

Reviewer 2: ```suggestion

### Initial taints (replaced), with out-of-tree controller

This approach uses `--register-with-taints` to apply a single initial taint at startup. A controller then atomically sets a set of replacement taints (configured using a custom resource) and removes the initial taint.

For each replacement taint, each component is then responsible for removing its own taint.

This is easier to maintain (no in-tree code) but requires people to run an additional
controller in their cluster.

```

Author: The out-of-tree controller option for managing the readiness-conditions are discussed as the next alternative. I added here with another top 'non-api' option that was considered - using 'prefixes' to identify readiness taints.

### Cluster-Scoped NodeReadinessRule CRD

An alternative design centered on a new CRD NodeReadinessRule, keeping the configuration outside of the node. This was rejected due to the risk of a single misconfigured CRD causing a large-scale cluster outage, making the node-local approach inherently safer.

CRD considered -

```yaml
apiVersion: readiness.k8s.io/v1
kind: NodeReadinessRule
metadata:
  name: networking-readiness
spec:
  nodeSelector:
    matchLabels:
      readiness-requirement: "network"
  requiredConditions:
  - type: "network.k8s.io/CalicoReady"
  - type: "network.k8s.io/NetworkProxyReady"

>   - type: "network.k8s.io/DRANetReady"

Reviewer 2: ```suggestion
  - type: "vendor.example/LowLatencyInterconnectReady"
```

  taint:
    key: "readiness.k8s.io/network-pending"
    effect: NoSchedule  
```

> 

Reviewer 2: How about a CRD that defines a set of rules that map (custom) conditions to taints?

Yes, you can break your cluster with a single, misguided cluster-scoped policy, but we already have that in other places (eg ValidatingAdmissionPolicy).

Author: Hi @lmktfy, thanks for the suggestion. Yes, I discussed this CRD but it was turned down last time due to the risk of misconfiguration. I developed a proof-of-concept on this here: https://github.com/ajaysundark/node-readiness-gate-controller, and following up with SIG-Node on this further. I'll update here in KEP with updates.

-  Non-preferred pods are scheduled onto the node before driver installation and readiness state is achieved, filling up valuable node capacity.

### Goals

The primary goals of this KEP are to:

1. Establish clear semantics using `NodeSpec` for declaring desired prerequisites (gates) and `NodeStatus` for reporting observed operational readiness (conditions) in node, aligning with Kubernetes API conventions.
1. Define a standard API field (`spec.readinessGates`) on the Node object to specify required readiness conditions beyond the default Kubelet `Ready` state.
1. Establish a standard pattern for gating a node's schedulability based on a set of required readiness conditions. The enforcement of these gates will be managed by a controller that applies a taint (e.g., `node.kubernetes.io/not-ready`) until all required conditions are met.
1. Centralize readiness reporting through a trusted, node-local probing mechanism (such as Kubelet or Node-Problem-Detector). This entity would be responsible for probing components and patching `node.status.Conditions`, which improves security by removing the need for multiple external agents to hold `nodes/status` patch permissions.

> 1. Improve scheduling correctness by also considering the reliability of the node lifecycle by preventing application pods from being scheduled onto nodes until declared readiness gates are met. 

Reviewer 3: ```suggestion
1. Improve scheduling correctness (via taints mechanism) by also considering the reliability of the node lifecycle by preventing application pods from being scheduled onto nodes until declared readiness gates are met. 
```

1. **Component Fails:** If there's a critical component rollout, and the readiness-probe detected it as health-check failure (based on thresholds and timeout configurations), it will update a blocking condition for node-scheduling.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

- API validation for the new `NodeSpec.readinessGates` field.
- Node Controller logic for taint-management based on `status.Conditions`.
- Verify readinessGates evaluation logic.

##### Integration tests

* Test node registration with gates defined.
* Test pod scheduling being blocked when gates are not met.
* Test pod scheduling succeeding when gates transition to `True`.
* Test existing scenarios with standard taints eg., `not-ready:NoExecute` for node timeout.

##### e2e tests

-  : 

### Graduation Criteria

* **Alpha:**
    * Feature implemented behind `NodeReadinessGates` feature gate, default `false`.
    * Basic unit and integration tests implemented.

>     * Initial API definition (`NodeSpec.readinessGates`) available.

Reviewer 4: do we have to start with the new field? Can we try out without the new field and see if approach works while in alpha?

* **Beta:**
    * Feature gate default `true`.
    * e2e tests implemented and passing consistently.

>     * Scalability and performance testing performed.

Reviewer 4: performance will mean some latency targets I would assume. Maybe say explicitly that we will check the latency
